src/ui/pages/work/partusage/PartUsageDialog.js

```diff
@@ -202,124 +239,231 @@ function PartUsageDialog(props) {
                 if (responseStoreCode) {
-                    showError(`Asset is already in a store (${responseStoreCode}).`);
+                    showError(`Asset "${assetIDCode}" is already in a store (${responseStoreCode}).`);
                     return undefined;
                 }
 
                 return assetData;
             }
 
             // Something is wrong if we get here (user input or code-related)
-            showError('Asset does not follow business rules.');
+            showError(`Asset "${assetIDCode}" does not follow business rules.`);
             return undefined;
 
         } catch (error){
             handleError(error);
         }
     };
 
-    const handleAssetSelectedWithPart = (assetData) => {
-        const { bin, partCode } = assetData;
+    const handleAssetSelectedWithPart = async (assetData) => {
+        const { bin, partCode, lot } = assetData;
+        const { transactionType, storeCode } = formData;
+
+        // When in an issue transaction, bin loading will also load the lot list
+        // since assets have a single bin (which triggers lot loading).
+        await Promise.all([
+            loadBinList(bin, partCode),
+            transactionType === returnType
+                ? loadLotList(transactionType, lot, bin, partCode, storeCode)
+                : null,
+        ]);
+    };
 
-        updateFormDataProperty(FORM.BIN, '');
-        if (assetData) {
-            updateFormDataProperty(FORM.BIN, bin);
-            loadBinList(bin, partCode);
-        }
+    const handleAssetDifferentOrNoPart = async (assetData) => {
+        const { partCode } = assetData;
+
+        // Clear the part field so the user is aware that the selected asset is associated with a different part
+        resetFieldWithDesc(FORM.PART, FORM.PART_DESC);
+
+        const results = await Promise.all([
+            loadPartData(partCode),
+            handleAssetSelectedWithPart(assetData),
+        ]);
+
+        const partData = results[0];
+        updateFormDataProperty(FORM.PART_DESC, partData?.description);
+        updateFormDataProperty(FORM.PART, partCode);
     };
 
-    const handleAssetUnspecifiedPart = async (assetData) => {
-        const { bin, partCode } = assetData;
+    const handleBinChange = async (bin) => {
+        const { transactionType, partCode, storeCode, assetIDCode, lot } = formData;
 
-        updateFormDataProperty(FORM.PART, '');
-        updateFormDataProperty(FORM.PART_DESC, '');
-        updateFormDataProperty(FORM.BIN, '');
-        if (assetData) {
-            updateFormDataProperty(FORM.BIN, bin);
-            loadBinList(bin, partCode);
-            updateFormDataProperty(FORM.PART, partCode);
+        // On a Return transaction, the lot can be filled before the bin (we expect there to be a lot list already),
+        // as such we must not clear the lot field and we can avoid re-loading the lot list.
+        if (transactionType === returnType) {
+            return;
+        }
```

Comment:
This is doing nothing, since the next if statement has the same body and the one after checks for a different transaction type. I would leave the comment though

Reply:
sure